Adaptation for static linking

miroi opened the pull request:
Hereby I am asking for the code review first (while doing the git rebase in between).

bast commented:
Hi Miro. Generally I think that you submit the PRs way too early. But I understand why.
The earlier the better for you since you get feedback earlier before all that coding.
The later the better for me since the product is more converged.
Difficult balance but in general please let the code first ripe on your fork
for few days/weeks to debug the corner cases. You don't need the code to immediately
enter the central repo thanks to distributed version control. You can develop multiple
features at the same time without the features entering immediately the central repo.
What is the central repo anyway in Git?

Comments:
- Shouldn't the `LINKING_STUFF` (please use a better name) be taken care of inside `math_libs.cmake`? Then it would be more transparent.
- Please describe what all the packages that we apt-get install in `.travis.yml` are for. New developers 5 years from now have to be able to understand why all those packages are installed. I don't know what all those packages are for.

miroi commented:
Hi Rado,

when my code is working on branch, and I foresee no substantial changes in the future, why waiting ? Both fresh and ripe codes will undergo review anyhow, and I don't believe that an 'old' code will have less referee comments than a 'fresh' one, as is this one.

Answer to comments:

i) I renamed LINKING_STUFF to MATH_LIBS_EXT in corresponding CMakeFiles.txt files. 
I am against putting the LINKING_STUFF -> MATH_LIBS_EXT variable into **math_libs.cmake** as for static linking it differs from compiler to compiler, from library to library (and even from version to version). Let the MATH_LIBS variable provides the very *core* for static linking, and remaining adaptations for various cases of static linking are to be done on user's side. Such is my experience with combination of Intel, PGI, GNU compilers / MKL,ATLAS, SYSTEM_NATIVE math libraries.

Even I was shaking with "holy awe" before modifying **math_libs.cmake** for static linking. Yes, I modified this file, in the minimalist and IMHO sufficient way.


ii) Packages added to .travis-ci.yml are aimed to ensure proper static linking on CI testing servers. Especially the **fc_lapack** static linking is non-trivial. After experimenting (how many commits were shot for that !) I had to resort to the static ATLAS library, and had switch linking order to "lapack-blas" in **math_libs.cmake**. One important case : only the binutils-gold package is able to ensure proper static linking on Linux (fortunately it exists on travis-ci Linux server(s), on my clusters I had to install it manually). 

With all these proposed changes for autocmake I am heading to the new static autocmake-DIRAC software to be used for multinode grid computing. I am sure the "autocmaked" DIRAC is better starting point in comparison to what I had before.

bast commented:
Hi Miro, just one thing about the waiting: You typically implement something that you need and you use. It is normal and typical that you discover bugs while using the new feature. It would be extremely atypical to code something that is just ready and bugfree and does not require refactoring or small or large changes. Therefore it is good to wait with the PR and use the code that you have written for some time before submitting it for a PR.

miroi commented:
Fine, I am going to perform all-combinations tests on this new feature and will display results in tabular form.

bast commented:
Hi Miro, you don't have to do it for me. I don't need any "executive summary". I just recommend
that you use the feature that you have implemented for few days before filing a PR (unless
it is a trivial change).

bast commented:
The test CMake sources should IMO be "thin". If all the static lib adaptations are in the test sources
but not in CMake modules that we tests, then I am not sure what we test in them. So either it can be done in a portable way and then it needs to be in a module or it cannot be done and then I think we should not even attempt it.

miroi commented:
Currently what is working is *static linking*  combination of  **gfortran+ATLAS/SYSTEM_NATIVE/MKL**, **ifort+ATLAS/SYSTEM_NATIVE/MKL**,  **pgf90+ATLAS/SYSTEM_NATIVE**.

The wanted (as prerequisite for static-DIRAC-for-grid) combination of **pgf90+MKL+static ** is not working due to this error:
~~~
ilias@login.grid.umb.sk:~/Work/qch/software/software_projects/autocmake_devel/autocmake_miroi/test/fc_blas/build_static_pgf90_mkl/.m
Linking Fortran executable ../bin/example
IPA: no IPA optimizations for 1 source files
/mnt/apps/intel/composer_xe_2013_sp1.1.106/mkl/lib/intel64/libmkl_core.a(mkl_aa_fw_load_orsl_lite_lib.o): In function `mkl_aa_fw_load_orsl_lite_lib':
../../../../serv/offload/framework/core/mkl_aa_fw_load_orsl_lite_lib.c:(.text+0xbe): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib64/libpthread.a(libpthread.o): In function `sem_open':
(.text+0x774d): warning: the use of `mktemp' is dangerous, better use `mkstemp'
/usr/bin/ld: dynamic STT_GNU_IFUNC symbol `strcmp' with pointer equality in `/usr/lib64/libc.a(strcmp.o)' can not be used when making an executable; recompile with -fPIE and relink with -pie
child process exit status 1: /usr/bin/ld
make[2]: *** [bin/example] Error 2
make[1]: *** [src/CMakeFiles/example.dir/all] Error 2
make: *** [all] Error 2
milias@login.grid.umb.sk:~/Work/qch/software/software_projects/autocmake_devel/autocmake_miroi/test/fc_blas/build_static_pgf90_mkl/.
~~~

I made quite heavy adaptation of the test's src/CMakeFiles.txt files . Maybe there is a space for moving some local cmake commands into the common  *static_linking.cmake* module. But first I would like to fix **pgf90+mkl+static** , later trying to make static tests thin.

Comments, please ?

bast commented:
Whatever works is fine. If this helps you achieving your goal then it is a good route.
Later you can refactor.

miroi commented on modules/math_libs.cmake:
```diff
@@ -49,6 +50,13 @@
 #-------------------------------------------------------------------------------
+# ENABLE_STATIC_LINKING
+
+if(ENABLE_STATIC_LINKING)
```
necessary (minor) change for static linking

miroi commented on modules/math_libs.cmake:
```diff
@@ -355,7 +363,8 @@ macro(config_math_service _SERVICE)
                 set(_omp_flag -fopenmp)
             endif()
-            if(MKL_COMPILER_BINDINGS MATCHES PGI)
+            # do not add -mp flag for PGI+MKL+STATIC_LINKING
```
Another change to achive PGI+MKL static linking.

miroi commented:
Hi,

finally I got PGI+MKL+static working !. I strived to keep minimum changes into modules/math_libs.cmake, so that the complexity of various compilers/libraries/linking options remains at user's src/CMakeFiles.txt side. There is no other way, IMHO.

All 9 combination of static linking (pgf90,ifort,gfortran/ATLAS,SYSTEM_NATIVE,MKL) is working (my "execution summary" on Linux cluster)!

I am missing the ACML library (3 more possibilities to check, maybe 6 with the ilp64 mode, or in total 12 when testing both dynamic and static linking) and the OpenBLAS library (another bunch of combinations for testing...). Also one should test it with IBM XL/ESSL, SYSTEM_NATIVE .... damn, another testing options...

So I am asking for this PR as I have serious plans with the static autocmake-DIRAC. The Slovak (empty) Grid environment is waiting.

miroi commented on test/fc_blas/src/CMakeLists.txt:
```diff
@@ -1,12 +1,31 @@
+endif()
+
 if(BLAS_FOUND)
-    if(CMAKE_Fortran_COMPILER_ID MATCHES PGI)
```
Who know what adaptations will come in for ACML, ESSL libraries, or for the IBM XL compiler ?

miroi commented on test/test.py:
```diff
@@ -156,3 +174,11 @@ def test_fc_lapack():
+
+def test_fc_lapack_static():
+    if sys.platform != 'win32':
+        stdout, stderr = configure_build_and_exe('fc_lapack', 'python setup.py --fc=gfortran --static --cmake-options="-DMATH_LIB_SEARCH_ORDER=ATLAS"')
```
The choice of ATLAS is necessary for static linking: the native lapack.a and blas.a are crashing due to repeating object, see https://github.com/scisoft/autocmake/issues/51#issuecomment-127579836 .

bast commented:
Sorry my changes force you to rebase and resolve but the conflicts should be trivial.

miroi commented:
Rebased.Any other changes?

bast commented:
Hi Miro, this is very important work. Thank you for that. The only thing that I don't like about
this change is that the tests carry all the workarounds and not the modules. This means that the tests
pass but if a user wants to use the module for static-linking it will then not work and one would need
to reintroduce the workarounds every time. Isn't there a way to hide those away in a module?
The test sources should ideally be free of hacks and workarounds otherwise there is IMO no point
for these tests. The tests now test themselves and not the modules.

miroi commented:
Fine,then Iet us wait,more iterations of this PR are to be done.I hope a lot of "hacks" will go to the mathlibs.cmake.

miroi commented on test/fc_lapack/src/CMakeLists.txt:
```diff
@@ -1,10 +1,15 @@
+    SET(CMAKE_SHARED_LIBRARY_LINK_Fortran_FLAGS)
+endif()
 
+if(CMAKE_Fortran_COMPILER_ID MATCHES Intel AND ENABLE_STATIC_LINKING)
```
Shall this remain in local src/CMakeListst.txt or should it go to compilers/COMPILER.Fortran.cmake  ?

miroi commented:
Comments please ?

bast commented:
I think this is good. We want the test sources to be thin. The reason is that we want users to be able
to plug CMake functionality with writing minimum CMake code on their side.

miroi commented:
Really good ? And concerning https://github.com/scisoft/autocmake/pull/54#discussion_r36615453 ?

miroi commented on test/fc_lapack/cmake/autocmake.cfg:
```diff
@@ -4,6 +4,9 @@ name: example
 [fc]
 source: ../../../modules/fc.cmake
 
+[cc]
```
This module should be removed, shouldn't ?

miroi commented:
Green tests; merge ?

bast commented:
Thanks Miro!
I am not 100% happy that we so heavily modify link flags that come from CMake
but perhaps we have to live with it for a while. Perhaps there is no other way.
I hope there is but we can explore it with the code in.
Why is the `[explicit]` block needed in `fc_blas` test?

miroi commented:
fixed and rebased

bast commented:
Hi Miro, can you please try one more thing:?
Currently we "brutally" erase all link flags with
```
set(CMAKE_SHARED_LIBRARY_LINK_Fortran_FLAGS)
```
Can you please try whether this works also:
```
list(REMOVE_ITEM CMAKE_SHARED_LIBRARY_LINK_Fortran_FLAGS "offending-flag")
```
This would be more gentle and more explicit.

miroi commented:
Hi,
works.
In general, for cases as this one I would prefer an automatic CDash testing. PGI/Intel/GNU/mathlibs/int64/static libs give too many combinations alltogether. We can not rely only on CI tests with free compilers and libs. Other projects, candidates for utilizing Autocmake, would  prefer the best compilers/libraries at hand, and these are usually commercial.

bast commented:
Thanks Miro. I have nothing against automated CDash testing in addition to the Travis/Appveyor CI.
But my ambition is to work towards a portable compiler-independent Autocmake which
can be tested by Travis/Appveyor.
